### PR TITLE
feat: load map pins from API endpoint

### PR DESCRIPTION
**Pull request by thisislawatts**
PR Checklist

- [ ] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

Switch to loading Map Pins from a REST API rather than hitting the Firestore database.

### CODE REVIEW

**Review by mariojsnunes — COMMENTED** [review]
Just to make sure... the pins we get from the API don't have all the info, so we need to load them again on click?

**Comment by mariojsnunes** [comment]
We can improve the `Maps` component logic if we simply use the url hash to manage the loaded pin state (we already do it on `showPinFromURL`.
`MapView` would navigate to the hash when a pin is clicked, and not export `onPinClicked` anymore.

**Comment by thisislawatts** [comment]
> Just to make sure... the pin list we get from the API doesn't have all the info, so we need to load them individually on click?

@mariojsnunes we are currently loading the map pin details from the user profile https://github.com/ONEARMY/community-platform/blob/156674420491c15f70d000ee9d5ea9a239bb0f44/src/stores/Maps/maps.store.ts#L142

Historically we used to store the information on the map pin under the `detail` property but at some point we dropped that in favour of having a single source of truth.

File: src/pages/Maps/Maps.tsx [hunk 1]
```diff
@@ -1,60 +1,85 @@
...
-      mapsStore.setActivePin(undefined)
+      setSelectedPin(null)
     }
   }, [])
 
-  useEffect(() => {
-    showPinFromURL()
-  }, [location.hash])
-
   const availableFilters = () => {
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply src/pages/Maps/Maps.tsx]: Addressed 3f820bd619ee7ceefe20ff9447cdf4ac4aaaa562

File: src/pages/Maps/Maps.tsx [hunk 2]
```diff
@@ -86,37 +111,56 @@ const MapsPage = observer(() => {
...
+   **/
   const showPinFromURL = async () => {
     const pinId = location.hash.slice(1)
-    // Only lookup if not already the active pin
-    if (pinId && pinId !== mapsStore.activePin?._id) {
-      const pin = await mapsStore.getPin(pinId)
-      if (pin._deleted) return
-      mapsStore.setActivePin(pin)
+    if (pinId) {
+      getPinByUserId(pinId)
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply src/pages/Maps/Maps.tsx]: Fixed: cef9c162166971b87189c30f57b6984120d43879

File: src/pages/Maps/Maps.tsx [hunk 3]
```diff
@@ -86,37 +111,56 @@ const MapsPage = observer(() => {
...
       />
       <MapView
+        activePin={selectedPin}
         mapRef={mapRef}
-        pins={filteredPins}
-        filters={activePinFilters}
-        onBoundingBoxChange={(boundingBox) =>
-          mapsStore.setMapBoundingBox(boundingBox)
-        }
+        pins={filterMapPinsByType(mapPins, activePinFilters)}
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply src/pages/Maps/Maps.tsx]: Fixed in 6d97f5b01d8ebec247647fa22fe2c7b6bde275ab

**Comment by thisislawatts** [comment]
> We can improve the `Maps` component logic if we simply use the url hash to manage the loaded pin state (we already do it on `showPinFromURL`). `MapView` would navigate to the hash when a pin is clicked, and not export `onPinClicked` anymore.

I've done some refactoring here so that we now use the `onPinClicked` method for mutating state. As a result the only component that knows we are using the location.hash to manage state is `Maps`. The aim of this change was to simplify MapView and centralise state in one place.

Generally across the application we use MobX or React for state management so we should aim to isolate deviation from this pattern (here it was using `window.location`) to a single component (Maps.tsx) in an effort to reduce maintenance costs. Before `location.hash` introduced non-obvious coupling between the two components.

File: src/pages/Maps/map.service.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+const getMapPins = async () => {
+  try {
+    const response = await fetch(API_URL + '/map-pins')
+    const mapPins = await response.json()
+
+    return mapPins
+  } catch (error) {
+    return []
```
Comment (mariojsnunes) [inline src/pages/Maps/map.service.ts:24]: should add some logging here

File: src/pages/Maps/map.service.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+const getMapPinByUserId = async (userName: string) => {
+  try {
+    const response = await fetch(API_URL + '/map-pins/' + userName)
+    const mapPin = await response.json()
+
+    return mapPin
+  } catch (error) {
+    logger.error('Failed to fetch map pins', error)
```
Comment (mariojsnunes) [inline src/pages/Maps/map.service.ts:35]: "Failed to fetch user pin"

File: src/pages/PageList.tsx [hunk 6]
```diff
@@ -104,7 +105,11 @@ const academy = {
 const maps = {
   moduleName: MODULE.MAP,
   path: '/map',
-  component: <MapsPage />,
+  component: (
+    <MapPinServiceContext.Provider value={mapPinService}>
```
Comment (mariojsnunes) [inline src/pages/PageList.tsx:109]: I'm curious why did you choose to implement `MapPinServiceContext`? For testability?

File: src/pages/Maps/Maps.tsx [hunk 7]
```diff
@@ -1,68 +1,92 @@
...
-    firstLoad: true,
-  })
+  }>(initialState)
+
+  const fetchMapPins = async () => {
+    const pins = await mapPinService.getMapPins()
+    setMapPins([...mapPins, ...pins])
+  }
+
+  const appendLoggedInUser = async (userName: string = '') => {
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply src/pages/Maps/Maps.tsx]: Resolved 89c2e0f9fe8de4f9e94adfa7404a703206df3fa7

File: src/pages/PageList.tsx (hunk 6 above)
Comment (thisislawatts) [reply src/pages/PageList.tsx:109]: Exactly that @mariojsnunes, it seemed like this was the cleanest way to introduce a mock for the service without having to use `jest.mock` which does some slightly magic things around hoisting.

> For this reason, Jest will automatically hoist jest.mock calls to the top of the module (before any imports).
https://jestjs.io/docs/manual-mocks#using-with-es-module-imports

When using `jest.mock` I also observed some interaction between it and the jest configuration inherited from react-scripts 🙈, so it seemed preferable to Keep It Simple and lean into the React hooks.

I am not tied to the pattern though if you have an alternative suggestion :)

**Review by mariojsnunes — APPROVED** [review]

**Pull request merged**